clients/scheduler/interval_action_test.go

```diff
@@ -0,0 +1,438 @@
+	Target:     "testName",
+	Protocol:   "321",
+	HTTPMethod: "post",
+	Address:    "127.0.0.1",
+	Port:       3000,
+	Path:       "321",
+	Publisher:  "321",
+	User:       "321",
+	Password:   "321",
+	Topic:      "321",
+}
+
+func TestIntervalActionRestClient_Add(t *testing.T) {
+	ts := testHttpServer(t, http.MethodPost, clients.ApiIntervalActionRoute)
+
+	defer ts.Close()
+
+	url := ts.URL + clients.ApiIntervalActionRoute
+
+	params := types.EndpointParams{
+		ServiceKey:  clients.SupportSchedulerServiceKey,
+		Path:        clients.ApiIntervalActionRoute,
+		UseRegistry: false,
+		Url:         url,
+		Interval:    clients.ClientMonitorDefault,
+	}
+
+	iac := NewIntervalActionClient(params, MockEndpoint{})
+
+	res, err := iac.Add(&testIntervalAction1, context.Background())
+
+	if res == "" {
+		t.Fatal("unexpected empty string response")
+	}
+
+	if err != nil {
+		t.Fatalf("unexpected error %s", err.Error())
+	}
+}
+
+func TestIntervalActionRestClient_Delete(t *testing.T) {
+	ts := testHttpServer(t, http.MethodDelete, clients.ApiIntervalActionRoute)
+
+	defer ts.Close()
+
+	url := ts.URL + clients.ApiIntervalActionRoute
+
+	params := types.EndpointParams{
+		ServiceKey:  clients.SupportSchedulerServiceKey,
+		Path:        clients.ApiIntervalActionRoute,
+		UseRegistry: false,
+		Url:         url,
+		Interval:    clients.ClientMonitorDefault,
+	}
+
+	ic := NewIntervalActionClient(params, MockEndpoint{})
+
+	err := ic.Delete(testID1, context.Background())
+
+	if err != nil {
+		t.Fatalf("unexpected error %s", err.Error())
+	}
+}
+
+func TestIntervalActionRestClient_DeleteByName(t *testing.T) {
+	ts := testHttpServer(t, http.MethodDelete, clients.ApiIntervalActionRoute)
+
+	defer ts.Close()
+
+	url := ts.URL + clients.ApiIntervalActionRoute
+
+	params := types.EndpointParams{
+		ServiceKey:  clients.SupportSchedulerServiceKey,
+		Path:        clients.ApiIntervalActionRoute,
+		UseRegistry: false,
+		Url:         url,
+		Interval:    clients.ClientMonitorDefault,
+	}
+
+	ic := NewIntervalActionClient(params, MockEndpoint{})
+
+	err := ic.DeleteByName(testIntervalAction1.Name, context.Background())
+
+	if err != nil {
+		t.Fatalf("unexpected error %s", err.Error())
+	}
+}
+
+func TestIntervalActionRestClient_IntervalAction(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+
+		if r.Method != http.MethodGet {
+			t.Fatalf("expected http method is GET, active http method is : %s", r.Method)
+		}
+
+		expectedURL := clients.ApiIntervalActionRoute + "/" + testID1
+		if r.URL.EscapedPath() != expectedURL {
+			t.Fatalf("expected uri path is %s, actual uri path is %s", expectedURL, r.URL.EscapedPath())
+		}
+
+		data, err := json.Marshal(testIntervalAction1)
+		if err != nil {
+			t.Fatalf("marshaling error: %s", err.Error())
+		}
+		w.Write(data)
+	}))
+
+	badJSONServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I am not a fan of starting a server to just be able to unit test a client. One reason is that is actually requires the test to obtain a port to listen on which is can possibly introduce issues depending on the environment. I think this is showing that the client should be a little more flexible and updated for better test-ability.

During the last round of comments I was suggesting to allow a caller to inject an "unmarshaler" and I think this still applies to this round of review. This would allow for easy testing, since you can inject an implementation which would just return an error and during happy path tests you can inject `json.Unmarshal` or some wrapper making these tests a lot simpler. This gives the added benefit of the client now being able to support many more content types(Open–closed principle).

Here is an example:

Create a type alias which will allow the caller to inject the unmarshaling function they wish to be applied. For example:

```go
type ResponseUnmarshaler func(data []byte, v interface{}) error
```

Then in your constructor(s), you can have the caller pass in that type:
```go

type intervalActionRestClient struct {
	urlClient interfaces.URLClient
        unmarshaler ResponseUnmarshaler
}

// NewIntervalActionClient creates an instance of IntervalActionClient
func NewIntervalActionClient(params types.EndpointParams, m interfaces.Endpointer, unmarshaler ResponseUnmarshaler) IntervalActionClient {
	return &intervalActionRestClient{urlClient: urlclient.New(params, m), unmarshaler: unmarshaler}
}
```

To support backwards compatibility you can keep the existing constructor and have it use `json.Unmarshal` as a default. This would work since all existing users of the construct are using it for `JSON` anyway or else it would not work.

Now, instead of creating a running server just to cover test cases where serialization has failed you can inject a custom `ResponseUnmarshaler` which always returns an `error`:
```go
// mockUnmarshalerError returns an error when unmarshaling is attempted.
func mockUnmarshalerError([]byte, interface{}) error {
	return errors.New("unmarshal error")
}

NewIntervalActionClient(params, endpointer, mockUnmarshalerError)
```

and the methods for `intervalActionRestClient` would refer to `intervalActionRestClient.unmarshaller` rather than `json.Unmarshal`:

```go
// Helper method to request and decode an interval action
func (iac *intervalActionRestClient) requestIntervalAction(
	urlSuffix string,
	ctx context.Context) (models.IntervalAction, error) {

	data, err := clients.GetRequest(urlSuffix, ctx, iac.urlClient)
	if err != nil {
		return models.IntervalAction{}, err
	}

	ia := models.IntervalAction{}
	err = ia.unmarshaler(data, &ia)
	if err != nil {
		return models.IntervalAction{}, err
	}

	return ia, nil
}
```

Furthermore, you can reuse the `ResponseUnmarshaler` type other packages which have similar structure to this allowing consistency in the project, so I think a good place for this is probably `request.go` to be used by all clients. All of this also applies to marshaling as well. An example of other clients having similar structure is:
https://github.com/edgexfoundry/go-mod-core-contracts/blob/4b1aba370a2a780425ec0b989b3dc7884f6a1289/clients/coredata/event.go#L102 


For more examples you can refer to the work in `go-mod-messaging` here is a good starting point. https://github.com/AnthonyMBonafide/go-mod-messaging/blob/a775bafe5b2419b22a509946a8df0703ec825d1e/internal/pkg/mqtt/client.go#L43-L69

Reply:
> I am not a fan of starting a server to just be able to unit test a client.

I concur.

> During the last round of comments I was suggesting to allow a caller to inject an "unmarshaler" and I think this still applies to this round of review.

Not sure why the original suggested direction wasn't incorporated in this latest round of changes.